lib/matplotlib/figure.py

```diff
@@ -1576,6 +1576,94 @@ def subplots(self, nrows=1, ncols=1, sharex=False, sharey=False,
             # Returned axis array will be always 2-d, even if nrows=ncols=1.
             return axarr
 
+    def build_grid(self, layout, subplot_kw=None, gridspec_kw=None):
+        """
+        Builds a layout based on an grid of strings
+
+        This is a helper function to build complex GridSpec layouts visually
+
+        Parameters
+        ----------
+        layout : list of list of str
+            A visual layout of how you want your Axes to be arranged labeled as strings.
+            For example ::
+
+               x = [['A', 'A', 'B'],
+                    ['C', 'D', 'B']]
```

Comment:
do you also want to support
```
x = ['AAB',
     'CDB']
```
?

Reply:
I suggested that and the argument @story645 had against it is that we should be encouraging long names like

```python
x = [['map', 'histogram'], 
     ['time_series', 'time_series']]
```

Reply:
but, that said, I suspect that could be implemented in the same pre-processing of the input that we would need to do for the auto-expanding logic.

Reply:
but that still works, the point is to support List[str] vs List[List[str]]
I don't have a really strong opinion there though

Reply:
if we support strings of arbitrary length, then how do we know that ["ABCD"] is ["A", "B", "C", "D"] vs ["AB", "CD"]?

Reply:
fair point.

Reply:
https://gist.github.com/thesamovar/52dbbb3a58a73c590d54c34f5f719bac has convinced me we should also support a _single_ string split on `\n` with no-nesting.

Reply:
What's the deterministic parse rule on that? Single letter equals 1 subplot?

Reply:
Yeah, you either pass in a single string _or_ a list.

```python
    @staticmethod
    def _normalize_grid_string(layout):
        return [list(ln) for ln in layout.strip().split('\n')]
```

Reply:
So the point I could be clearer in making is the docs  have to be 100 percent explicit that for string input, each character (excluding new lines and , ) is parsed as a subplot.

Reply:
First of all - thanks for looking at that gist - would love to see this in matplotlib. I'd definitely encourage you to include the multi-line string approach because even though it's not super-generalisable it covers a very common use case for scientists preparing figures.

Two other random notes:
1. I updated my gist so that you can now use a '.' to indicate an empty axis because then I don't have to worry about spaces.
2. If you can find a good way to add the panel labels so that they align correctly with ``tight_layout`` even with different axis sizes without having to do any fiddling, that would be amazing. There might already be a good way of doing this but I've never figured it out. My current solution in the gist sort of works but is a bit hacky and the alignment isn't quite right (the letters should line up on the left not the right).

Thanks for all your work on matplotlib - it's very much appreciated.

Reply:
[constrained_layout](https://matplotlib.org/3.1.1/tutorials/intermediate/constrainedlayout_guide.html) handles empty gridspec slots by placing invisible axes in the empty slots.